Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorCore.cs

```
@@ -75,23 +80,99 @@ public override async Task StopAsync()
+            IReadOnlyList<DocumentServiceLease> documentServiceLeases = await this.documentServiceLeaseStoreManager
+                .LeaseContainer
+                .GetAllLeasesAsync()
+                .ConfigureAwait(false);
+
+            // No lease documents. Return.
+
+            if (documentServiceLeases.Count == 0)
+            {
+                return;
+            }
+
+            DocumentServiceLease documentServiceLease = documentServiceLeases.FirstOrDefault();
+
+            // No lease documents that match the Id.
+
+            if (documentServiceLease == default)
+            {
+                return;
+            }
+
+            // Mode attribute exists on lease document, but it is not set. legacy is always LatestVersion because
+            // AllVersionsAndDeletes does not exist. There should not be any legacy lease documents that are
+            // AllVersionsAndDeletes. If the ChangeFeedProcessor's mode is not legacy, a CosmosException should thrown.
+            // If the ChangeFeedProcessor mode is not the mode in the lease document, a CosmosException should be thrown.
+
+            bool shouldThrowException = string.IsNullOrEmpty(documentServiceLease.Mode)
+                ? this.VerifyChangeFeedProcessorMode(
+                    changeFeedMode: ChangeFeedMode.LatestVersion,
+                    leaseChangeFeedMode: documentServiceLease.Mode,
+                    normalizedProcessorChangeFeedMode: out string normalizedProcessorChangeFeedMode)
+                : this.VerifyChangeFeedProcessorMode(
+                    changeFeedMode: this.changeFeedLeaseOptions.Mode,
+                    leaseChangeFeedMode: documentServiceLease.Mode,
+                    normalizedProcessorChangeFeedMode: out normalizedProcessorChangeFeedMode);
+
+            // If shouldThrowException is true, throw the CosmosException.
+
+            if (shouldThrowException)
+            {
+                throw CosmosExceptionFactory.CreateBadRequestException(
```

Should we throw a CosmosException with Status Code HTTP 400 or simply an InvalidArgumentException or InvalidOperationException? 

CosmosException might be useful mostly in cases where either we represent a service response or the customer expects a service response from the operation. But in this case, this is CFP.StartAsync, it has no return type, and customers do not expect any service response.

Thoughts @kirankumarkolli ?

---

I could rock with the `InvalidArgumentException` personally. The use of the `CosmosException` seemed a bit much I suppose. And defaulting things like header because we don't have it was not ideal. So, I vote for `InvalidArgumentException`. But we can wait to see what @kirankumarkolli thinks if we must. Also, need another reviewer if Kiran does not mind.

---

If I don't get a weigh in, I will make the decision to change to `InvalidArgumentException` and we can always open this up for discussion later.

---

And by `InvalidArgumentException` we really mean `ArgumentException` just to clarify.